### PR TITLE
feat: replace CLS with ALS, enable ALS by default, discourage unmanaged transactions

### LINKED ISSUE #10952: removeIndex not working with schema
## What are you doing?

```js
module.exports = {
  up: (queryInterface, Sequelize) => {
    return queryInterface.addIndex({
      tableName: 'user',
      schema: 'core',
    }, [
      'username',
    ]);
  },

  down: (queryInterface, Sequelize) => {
    return queryInterface.removeIndex({
      tableName: 'user',
      schema: 'core',
    }, [
      'username',
    ]);
  },
};
```
**To Reproduce**
1. run migraion
2. undo migration

## What do you expect to happen?
add and remove index

## What is actually happening?
index not removed 

the query generated in wrong
```sql
 DROP INDEX IF EXISTS "[object _object]_username"
```

## Environment
Dialect:
- [ ] mysql
- [x] postgres
- [ ] sqlite
- [ ] mssql
- [ ] any
Dialect **library** version:  7.11.0
Database version: PostgreSQL 11.2, compiled by Visual C++ build 1914, 64-bit
Sequelize version: 5.8.6
Node Version: 10.15.3
OS: Windows 10
If TypeScript related: TypeScript version: XXX
Tested with latest release:
- [ ] No
- [ ] Yes, specify that version:

### LINKED ISSUE #13742: Add type for Sequelize.useCLS parameter
## Issue Creation Checklist

[x] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

## Feature Description

A type for the `Sequelize.useCLS` would codify the true requirements that Sequelize has for a namespace to act as CLS for continuation-local transaction storage.

### Is your feature request related to a problem? Please describe.

I've implemented a small class that wraps `AsyncLocalStorage`, since we're on node 16. Figuring out the precise requirements/expectations of this object has required a lot of digging into sequelize internals.

### Describe the solution you'd like

It would be great to have an interface with code comments for each method that is needed and what is should do, so that sequelize can use one's object for transaction CLS. Here's a list of requirements I've gathered by digging around:

- A `run` method to call a block in scope where the storage is accessible
- `get` and `set` methods that accept a `string` key parameter and a value `Transaction | null` parameter
- A bind method, which I've left unimplemented—doesn't seem to be used (but it's enforced in the called to `Sequelize.useCLS`). I'm guessing this was added as a safety check to be extra sure someone was providing a "real" namespace from cls-hooked. Maybe this requirement can be removed? Or someone can correct me?

### Why should this be in Sequelize

`Sequelize.useCLS` allows providing external transaction CLS. Codifying/documenting the interface here would empower others to bring whatever the latest/greatest CLS mechanism is, without Sequelize needing (hopefully) to add explicit support for storage X or storage Y. This seems especially useful, as we're coming up on the 2 year anniversary of the last time [cls-hooked](https://github.com/Jeff-Lewis/cls-hooked) was updated (latest commit as of now is Dec 8, 2019).

### Describe alternatives/workarounds you've considered

Right now, we're getting by just reading sequelize internals and implementing methods that abide by the implied interface.

### Additional context

Happy to add more, but I think I've added everything.

### Is this feature dialect-specific?

- [x] No. This feature is relevant to Sequelize as a whole.
- [ ] Yes. This feature only applies to the following dialect(s): XXX, YYY, ZZZ

### Would you be willing to implement this feature by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [ ] Yes, I have the time but I don't know how to start, I would need guidance.
- [x] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

### LINKED ISSUE #15291: Migrate CLS to AsyncLocalStorage
CLS is outdated and Node has supported a built-in AsyncLocalStorage for long enough now. It's time to drop CLS.

Let's also use this refactor as an opportunity for making ALS an option that can be configured per Sequelize instance.

This change should make #13742 redundant.

### PR DESCRIPTION
**Pull request by ephys**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

Closes #15291
Closes #13742

This PR drops CLS in favor of using node's built-in AsyncLocalStorage.

The AsyncLocalStorage is now *per Sequelize instance* instead of global. It never made sense for this option to be global.

It also enables ALS transactions by default. They can be disabled by setting `disableAlsTransactions` to true in Sequelize's options.

Finally, in order to discourage using unmanaged transactions (which are easy to misuse), this PR removes the ability to get an unmanaged transaction by not providing a callback to `sequelize.transaction()`.
Unmanaged transactions are now started using `sequelize.startUnmanagedTransaction()` instead.

## Todos

- [ ] Open a PR to update the documentation website

### CODE REVIEW

File: src/sequelize-typescript.ts [hunk 1]
```diff
@@ -170,4 +175,148 @@ export class SequelizeTypeScript {
...
+   * We highly recommend using {@link Sequelize#transaction} instead.
+   * If you really want to use the manual solution, don't forget to commit or rollback your transaction once you are done with it.
+   *
+   * Transactions started by this method are not automatically passed to queries. You must pass the transaction object manually,
+   * even if {@link Options.disableAlsTransactions} is false.
+   *
+   * @example
+   * ```ts
+   * try {
+   *   const transaction = await sequelize.unsafeStartManualTransaction();
```
Comment (1valdis) [inline src/sequelize-typescript.ts:300]:
```suggestion
   *   const transaction = await sequelize.startUnmanagedTransaction();
```

File: test/integration/als.test.ts [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+          }
+        }, 200);
+      });
+      expect(transactionEnded).not.to.be.ok;
+
+      expect(vars.alsSequelize.getCurrentAlsTransaction()).not.to.be.ok;
+
+      // Just to make sure it didn't change between our last check and the assertion
+      expect(transactionEnded).not.to.be.ok;
+      await clsTask; // ensure we don't leak the promise
```
Comment (WikiRik) [inline test/integration/als.test.ts:105]:
```suggestion
      await alsTask; // ensure we don't leak the promise
```

**Review by WikiRik — COMMENTED** [review]
Looks good! Don't forget to remove `cls-hooked` from package.json, apart from that it's just a few typos

File: test/integration/als.test.ts [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    });
+
+    it('does not leak variables to the outer scope', async () => {
+      // This is a little tricky. We want to check the values in the outer scope, when the transaction has been successfully set up, but before it has been comitted.
+      // We can't just call another function from inside that transaction, since that would transfer the context to that function - exactly what we are trying to prevent;
+
+      let transactionSetup = false;
+      let transactionEnded = false;
+
+      const clsTask = vars.alsSequelize.transaction(async () => {
```
[the remark this answers is not in the document]
Comment (ephys) [reply test/integration/als.test.ts:83]: Done

File: test/integration/als.test.ts (hunk 2 above)
Comment (ephys) [reply test/integration/als.test.ts]: Done

File: src/sequelize-typescript.ts (hunk 1 above)
Comment (ephys) [reply src/sequelize-typescript.ts]: Thank you for reviewing :) Done

**Review by WikiRik — COMMENTED** [review]
I found a last reference to CLS here; https://github.com/sequelize/sequelize/blob/a2df9799b1ecd6d3599eaa6784d2e380a16c2e0e/test/unit/dialects/abstract/query.test.js#L478
But we should be good after that

**Comment by ephys** [comment]
I don't know why that's called CLS but it doesn't seem to be related to Continuation local storage

**Comment by WikiRik** [comment]
Maybe rename it to `customQuery`?

**Review by WikiRik — APPROVED** [review]
We can do the rename in a different PR

File: src/sequelize-typescript.ts (hunk 1 above)
Comment (1valdis) [reply src/sequelize-typescript.ts]: I was just curious what's going to happen to unmaintained `cls-hooked` package and managed transactions in Sequelize, and went across this PR😅

**Comment by ephys** [comment]
Yes let's leave it as is, we'll update the test suite when we migrate it and/or query.js to TypeScript

**Comment by OmriSama** [comment]
The `v7` alpha documentation still mentions `cls-hooked`; is it going to be updated?

**Comment by ephys** [comment]
It is, the update is in https://github.com/sequelize/website/pull/419, which I need to update & get merged

**Pull request merged**